### ISSUE TITLE
AOT autograd graph recompilation for constant_pad with dynamic shape 

### ISSUE THREAD

**Issue by sujoysaraswati** [pytorch/pytorch#97348]
### 🐛 Describe the bug

The following test case recompiles with AOT autograd even with dynamic=True and dynamo.config.specialize_int = False

```
import torch.nn.functional as F
import torch
import torch._dynamo as dynamo
from torch._functorch.aot_autograd import aot_module_simplified

dynamo.config.specialize_int = False
dynamo.reset()

def my_aot_compiler(gm, example_inputs):
    def my_compiler(gm, example_inputs):
        print(gm.code)
        return gm.forward

    # Invoke AOTAutograd
    return aot_module_simplified(
        gm,
        example_inputs,
        fw_compiler=my_compiler
    )

def my_example(t, p):
  m = torch.nn.ConstantPad2d(p, 3.5)
  out = m(t)
  return out   

compiled_fn = torch.compile(backend=my_aot_compiler, dynamic=True)(my_example)

t = torch.empty(2, 3, 4).requires_grad_(True)
p = (1, 1, 0, 0)
r = compiled_fn(t, p)
print(f"Output (1st) size = {r.size()}")

t2 = torch.empty(2, 3, 4).requires_grad_(True)
p2 = (2, 2, 0, 0)
r2 = compiled_fn(t2, p2)
print(f"Output (2nd) size = {r2.size()}")

t3 = torch.empty(2, 3, 4).requires_grad_(True)
p3 = (3, 3, 0, 0)
r3 = compiled_fn(t3, p3)
print(f"Output (3rd) size = {r3.size()}")
```
The output graphs seems to sometimes specialize the ints (first call), and sometimes doesn't. Even when it doesn't specialize the ints, it recompiles the graph (second and third calls) -

```
def forward(self, primals_1):
    constant_pad_nd = torch.ops.aten.constant_pad_nd.default(primals_1, [1, 1, 0, 0], 3.5);  primals_1 = None
    return [constant_pad_nd]
    
Output (1st) size = torch.Size([2, 3, 6])



def forward(self, primals_1, primals_2, primals_3):
    constant_pad_nd = torch.ops.aten.constant_pad_nd.default(primals_1, [primals_2, primals_3, 0, 0], 3.5);  primals_1 = None
    return [constant_pad_nd, primals_3, primals_2]
    
Output (2nd) size = torch.Size([2, 3, 8])
/usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:1216: UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
/usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:1216: UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(



def forward(self, primals_1, primals_2, primals_3):
    constant_pad_nd = torch.ops.aten.constant_pad_nd.default(primals_1, [primals_2, primals_3, 0, 0], 3.5);  primals_1 = None
    return [constant_pad_nd, primals_2, primals_3]
    
Output (3rd) size = torch.Size([2, 3, 10])
/usr/local/lib/python3.9/dist-packages/torch/_functorch/aot_autograd.py:1216: UserWarning: Your compiler for AOTAutograd is returning a function that doesn't take boxed arguments. Please wrap it with functorch.compile.make_boxed_func or handle the boxed arguments yourself. See https://github.com/pytorch/pytorch/pull/83137#issuecomment-1211320670 for rationale.
  warnings.warn(
```

### Versions

PyTorch version: 2.1.0.dev20230322+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.5 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: 10.0.0-4ubuntu1 
CMake version: version 3.26.0
Libc version: glibc-2.31

Python version: 3.9.16 (main, Dec  7 2022, 01:11:51)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-5.10.147+-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.8.89
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Tesla T4
Nvidia driver version: 525.85.12
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.7.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.7.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   46 bits physical, 48 bits virtual
CPU(s):                          2
On-line CPU(s) list:             0,1
Thread(s) per core:              2
Core(s) per socket:              1
Socket(s):                       1
NUMA node(s):                    1
Vendor ID:                       GenuineIntel
CPU family:                      6
Model:                           85
Model name:                      Intel(R) Xeon(R) CPU @ 2.00GHz
Stepping:                        3
CPU MHz:                         2000.198
BogoMIPS:                        4000.39
Hypervisor vendor:               KVM
Virtualization type:             full
L1d cache:                       32 KiB
L1i cache:                       32 KiB
L2 cache:                        1 MiB
L3 cache:                        38.5 MiB
NUMA node0 CPU(s):               0,1
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Mitigation; PTE Inversion
Vulnerability Mds:               Vulnerable; SMT Host state unknown
Vulnerability Meltdown:          Vulnerable
Vulnerability Mmio stale data:   Vulnerable
Vulnerability Retbleed:          Vulnerable
Vulnerability Spec store bypass: Vulnerable
Vulnerability Spectre v1:        Vulnerable: __user pointer sanitization and usercopy barriers only; no swapgs barriers
Vulnerability Spectre v2:        Vulnerable, IBPB: disabled, STIBP: disabled, PBRSB-eIBRS: Not affected
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Vulnerable
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat md_clear arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-triton==2.1.0+e650d3708b
[pip3] torch==2.1.0.dev20230322+cu117
[pip3] torchaudio==0.13.1+cu116
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.14.1
[pip3] torchvision==0.14.1+cu116
[conda] Could not collect

cc @ezyang @soumith @msaroufim @wconstab @ngimel @bdhirsh @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

**Comment by SherlockNoMad** [comment]
cc @ezyang, @Chillee, @voznesenskym

**Comment by ezyang** [comment]
Yes, unfortunately inductor specialized in constant pad lowering atm

**Comment by ngimel** [comment]
Not anymore! #97099

**Comment by ezyang** [comment]
Oh oops, I was too fast. But it is recompiling between 2 and 3

The real problem is overaggressive guard on `(2, 2, 0, 0)`. This is the same as https://github.com/pytorch/pytorch/issues/93344 cc @yanboliang

**Comment by yanboliang** [comment]
The root cause is the same as #93344, as dynamo specialize int list:
https://github.com/pytorch/pytorch/blob/416bac5b813a181753afade781ae30f4f0843586/torch/_dynamo/variables/builder.py#L585-L590

The decision was made based on the pattern on TB/HF/TIMM models, where they expect the int list are constant. However, since we saw more and more cases don't follow this pattern, I proposed to add a config ```specialize_int_list``` to support turning this off if users don't expect in list are constant. I can send a PR if there is no objection. cc @ezyang @ngimel @SherlockNoMad
